PushSharp.Apple/AppleHttpPushChannel.cs

```diff
@@ -0,0 +1,210 @@
+                                {
+                                    IsSubscriptionExpired = true,
+                                    OldSubscriptionId = appleNotification.DeviceToken
+                                };
+                            }
+                            else
+                            {
+                                result = new SendNotificationResult(notification, false);
+                            }
+                            callback(this, result);
+                            return;
+                        }
+                    }
+
+                    var retryNotification = true;
+                    // this response are non recoverable, do not retry the notification
+                    switch (response.StatusCode)
+                    {
+                        case HttpStatusCode.Forbidden:
+                        case HttpStatusCode.BadRequest:
+                        case HttpStatusCode.NotFound:
+                        case HttpStatusCode.MethodNotAllowed:
+                        case HttpStatusCode.RequestEntityTooLarge:
+                            retryNotification = false;
+                            break;
+                        default:
+                            break;
+                    }
+
+                    if (callback != null)
+                    {
+                        callback(this, new SendNotificationResult(notification, retryNotification, new Exception("Error during APNS Send.")));
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Error("Exception during APNS Send with channel {2}: {0} -> {1}",
+                    appleNotification.Identifier,
+                    ex,
+                    _channelInstanceId);
+
+                if (callback != null)
+                {
+                    // will requeue the notification
+                    callback(this, new SendNotificationResult(notification, true, ex));
+                }
+            }
+        }
+
+        private string GetJwtToken()
+        {
+            lock (_lock)
+            {
+                if (!string.IsNullOrEmpty(_currentJtw))
+                {
+                    // checks expiration (50m)
+                    var tokenExpiration = DateTime.UtcNow - _jtwCreationDate.Value;
+                    if (tokenExpiration.TotalMinutes < 50)
+                        return _currentJtw;
+                }
+
+                // create a new token
+                var dateTimeUtcNow = DateTime.UtcNow;
+                var header = JsonConvert.SerializeObject(new { alg = "ES256", kid = _appleSettings.KeyId });
+                var payload = JsonConvert.SerializeObject(new { iss = _appleSettings.TeamId, iat = ToEpoch(dateTimeUtcNow) });
+
+                var key = CngKey.Import(Convert.FromBase64String(_appleSettings.PrivateKey), CngKeyBlobFormat.Pkcs8PrivateBlob);
```

Comment:
Do we need to make a new key every time?

Reply:
not needed, moved to constructor